src/main/java/org/fcrepo/migration/handlers/ocfl/ArchiveGroupHandler.java

```diff
@@ -255,15 +255,22 @@ private void writeDatastreamContent(final DatastreamVersion dv,
         final var datastreamId = dv.getDatastreamInfo().getDatastreamId();
         final var datastreamControlGroup = dv.getDatastreamInfo().getControlGroup();
         if (fedora3DigestValid(f3Digest)) {
-            try (var digestStream = new DigestInputStream(contentStream,
-                    MessageDigest.getInstance(f3Digest.getType()))) {
-                session.writeResource(datastreamHeaders, digestStream);
-                final var expectedDigest = f3Digest.getDigest();
-                final var actualDigest = Bytes.wrap(digestStream.getMessageDigest().digest()).encodeHex();
-                if (!actualDigest.equalsIgnoreCase(expectedDigest)) {
-                    final var msg = String.format("%s/%s: digest %s doesn't match expected digest %s",
-                            ocflObjectId, datastreamId, actualDigest, expectedDigest);
-                    throw new RuntimeException(msg);
+            try {
+                final var messageDigest = MessageDigest.getInstance(f3Digest.getType());
+                if (session instanceof PlainOcflObjectSession) {
+                    session.writeResource(datastreamHeaders, contentStream);
+                    ((PlainOcflObjectSession) session).addDigest(datastreamHeaders, f3Digest);
```

Comment:
The digest should be added to the resource headers on line 437. Would it work to read that value in the `writeResource` call rather than introducing this method?

Reply:
One possible issue with that is that I have to be able to turn off the checking - so if I'm adding the digest in writeResource, I have to be able to tell the session that we're not checking the fedora 3 checksums, so it shouldn't add the digest...

Reply:
Could you perhaps toggle it in the PlainOcflObjectSession when it's created?

Reply:
sure, let me give that a try

Reply:
I added a commit - is that what you had in mind, or something else?

Reply:
Yes, that's what I had in mind. Can you change the if statement on line 260 to check the `migrationType` rather than doing the `instanceof` check?

Reply:
done